repl/src/main/scala/ammonite/repl/interp/Interpreter.scala

```diff
@@ -311,14 +349,66 @@ class Interpreter(prompt0: Ref[String],
       (classFiles, newImports) <- compiled
       _ = Timer("cachedCompileBlock 4")
       cls <- eval.loadClass(fullyQualifiedName, classFiles)
-    } yield (cls, newImports)
+    } yield (cls, newImports, tag)
+
   }
 
-  def processModule(source: ImportHook.Source, code: String, wrapperName: Name, pkgName: Seq[Name]) = {
 
-    processModule0(source, code, wrapperName, pkgName, predefImports)
+  def evalCachedClassFiles(cachedData: Seq[CompileCache], pkg: String, wrapper: String): Unit = {
+
+    def evalMain(cls: Class[_]) =
+      cls.getDeclaredMethod("$main").invoke(null)
+
+    var blockNumber = 1;
+    def getBlockNumber = blockNumber match {
+      case 1 => ""
+      case _ => "_" + blockNumber.toString
+    }
+
+    // Iterate through cached blocks and execute classFiles
+    // blockNumber keeps track of blockIndex
+    cachedData.foreach { d =>
+      for {
+        cls <- eval.loadClass(pkg + "." + wrapper + getBlockNumber, d._1)
```

Comment:
What happens if `loadClass` or `evalMain` fail? You should probably use `Res.map` on this, and make `evalCachedClassFiles` return a `Res[_]` to represent the possibility of failure